File: dotnet/DominoFactory/Measurement.cs
// <copyright file="Measurement.cs">Copyright (c) Peter Rosser. All rights reserved.</copyright>

namespace ShaperDominoLib;

using System;
using System.Globalization;

/// <summary>
/// Represents a measurement value with an associated unit of measure.
/// </summary>
public readonly record struct Measurement(double Value, UnitOfMeasure Unit)
{
    private const double MillimetersPerInch = 25.4;

    /// <summary>
    /// Parses a string representation of a measurement.
    /// Supports: "8.5in", "8 1/2\"", "8-1/2in", "210mm", "21cm", "1.5" (assumes inches).
    /// Imperial measurements support fractions. Metric does not.
    /// </summary>
    /// <param name="s">The string to parse.</param>
    /// <returns>A new Measurement instance.</returns>
    /// <exception cref="FormatException">Thrown when the string format is invalid.</exception>
    public static Measurement Parse(string s)
    {
        if (string.IsNullOrWhiteSpace(s))
        {
            throw new FormatException("Measurement string cannot be null or empty.");
        }

        s = s.Trim();

        // Check for metric unit suffixes (cm, mm)
        if (s.EndsWith("cm", StringComparison.OrdinalIgnoreCase))
        {
            string valueStr = s[..^2].Trim();
            if (double.TryParse(valueStr, NumberStyles.Float, CultureInfo.InvariantCulture, out double value))
            {
                return (value * 10).Millimeters(); // Convert cm to mm
            }

            throw new FormatException($"Invalid centimeter measurement: '{s}'.");
        }
        else if (s.EndsWith("mm", StringComparison.OrdinalIgnoreCase))
        {
            string valueStr = s[..^2].Trim();
            return double.TryParse(valueStr, NumberStyles.Float, CultureInfo.InvariantCulture, out double value)
                ? value.Millimeters()
                : throw new FormatException($"Invalid millimeter measurement: '{s}'.");
        }

        // Check for imperial unit suffixes (in, ")
        string imperialValueStr;
        if (s.EndsWith("in", StringComparison.OrdinalIgnoreCase))
        {
            imperialValueStr = s[..^2].Trim();
        }
        else if (s.EndsWith('"'))
        {
            imperialValueStr = s[..^1].Trim();
        }
        else
        {
            // No unit specified, assume inches
            imperialValueStr = s;
        }

        // Try parsing imperial with fraction support
        return TryParseImperialWithFraction(imperialValueStr, out double inches)
            ? inches.Inches()
            : throw new FormatException($"Invalid measurement format: '{s}'. Expected format: '8.5', '8 1/2\"', '8-1/2in', '210mm', or '21cm'.");
    }

    private static bool TryParseImperialWithFraction(string s, out double result)
    {
        s = s.Trim();

        // Try simple decimal first
        if (double.TryParse(s, NumberStyles.Float, CultureInfo.InvariantCulture, out result))
        {
            return true;
        }

        // Check for fractions: supports "8 1/2" or "8-1/2" (proper) and "17/2" (improper)
        int wholeNumber = 0;
        string fractionPart = s;

        // Handle proper fractions with space or dash separator (e.g., "8 1/2" or "8-1/2")
        int spaceIndex = s.IndexOf(' ');
        int dashIndex = s.IndexOf('-');
        int separatorIndex = spaceIndex >= 0 ? spaceIndex : dashIndex;

        if (separatorIndex > 0)
        {
            string wholePart = s[..separatorIndex].Trim();
            if (!int.TryParse(wholePart, NumberStyles.Integer, CultureInfo.InvariantCulture, out wholeNumber))
            {
                return false;
            }

            fractionPart = s[(separatorIndex + 1)..].Trim();
        }

        // Parse the fraction part (e.g., "1/2")
        int slashIndex = fractionPart.IndexOf('/');
        if (slashIndex <= 0)
        {
            return false; // No fraction found
        }

        string numeratorStr = fractionPart[..slashIndex].Trim();
        string denominatorStr = fractionPart[(slashIndex + 1)..].Trim();

        if (!int.TryParse(numeratorStr, NumberStyles.Integer, CultureInfo.InvariantCulture, out int numerator) ||
            !int.TryParse(denominatorStr, NumberStyles.Integer, CultureInfo.InvariantCulture, out int denominator) ||
            denominator == 0)
        {
            return false;
        }

        result = wholeNumber + (double)numerator / denominator;
        return true;
    }

    /// <summary>
    /// Tries to parse a string representation of a measurement.
    /// </summary>
    /// <param name="s">The string to parse.</param>
    /// <param name="result">The parsed measurement if successful.</param>
    /// <returns>True if parsing succeeded, false otherwise.</returns>
    public static bool TryParse(string s, out Measurement result)
    {
        try
        {
            result = Parse(s);
            return true;
        }
        catch
        {
            result = default;
            return false;
        }
    }

    /// <summary>
    /// Converts this measurement to inches.
    /// </summary>
    /// <returns>The value in inches.</returns>
    public double ToInches()
    {
        return Unit == UnitOfMeasure.Inches ? Value : Value / MillimetersPerInch;
    }

    /// <summary>
    /// Converts this measurement to millimeters.
    /// </summary>
    /// <returns>The value in millimeters.</returns>
    public double ToMillimeters()
    {
        return Unit == UnitOfMeasure.Millimeters ? Value : Value * MillimetersPerInch;
    }

    /// <summary>
    /// Returns a string representation of this measurement.
    /// </summary>
    /// <returns>A string in the format "value unit".</returns>
    public override string ToString()
    {
        return Unit == UnitOfMeasure.Inches ? $"{Value:F3}in" : $"{Value:F2}mm";
    }
}

File: dominoes-web-app/src/lib/shaper-domino/measurement.ts
import type { UnitOfMeasure } from './unitOfMeasure';
import { UnitOfMeasure as Units } from './unitOfMeasure';

const MILLIMETERS_PER_INCH = 25.4;

export class Measurement {
	readonly value: number;
	readonly unit: UnitOfMeasure;

	constructor(value: number, unit: UnitOfMeasure) {
		if (!Number.isFinite(value)) {
			throw new Error('Measurement value must be a finite number.');
		}
		this.value = value;
		this.unit = unit;
	}

	static parse(input: string): Measurement {
		if (!input || !input.trim()) {
			throw new Error('Measurement string cannot be empty.');
		}

		const s = input.trim();

		// Metric suffixes
		if (s.toLowerCase().endsWith('cm')) {
			const valueStr = s.slice(0, -2).trim();
			const value = Number.parseFloat(valueStr);
			if (!Number.isFinite(value)) {
				throw new Error(`Invalid centimeter measurement: '${input}'.`);
			}
			return new Measurement(value * 10, Units.Millimeters);
		}

		if (s.toLowerCase().endsWith('mm')) {
			const valueStr = s.slice(0, -2).trim();
			const value = Number.parseFloat(valueStr);
			if (!Number.isFinite(value)) {
				throw new Error(`Invalid millimeter measurement: '${input}'.`);
			}
			return new Measurement(value, Units.Millimeters);
		}

		// Imperial suffixes: in or "
		let imperialPart = s;
		if (s.toLowerCase().endsWith('in')) {
			imperialPart = s.slice(0, -2).trim();
		} else if (s.endsWith('"')) {
			imperialPart = s.slice(0, -1).trim();
		}

		const inches = tryParseImperialWithFraction(imperialPart);
		if (inches == null) {
			throw new Error(
				`Invalid measurement format: '${input}'. Expected: '8.5', '8 1/2"', '8-1/2in', '210mm', or '21cm'.`
			);
		}

		return new Measurement(inches, Units.Inches);
	}

	static tryParse(input: string): Measurement | null {
		try {
			return Measurement.parse(input);
		} catch {
			return null;
		}
	}

	toInches(): number {
		return this.unit === Units.Inches ? this.value : this.value / MILLIMETERS_PER_INCH;
	}

	toMillimeters(): number {
		return this.unit === Units.Millimeters ? this.value : this.value * MILLIMETERS_PER_INCH;
	}

	toString(): string {
		if (this.unit === Units.Inches) {
			return `${this.value.toFixed(3)}in`;
		}
		return `${this.value.toFixed(2)}mm`;
	}
}

function tryParseImperialWithFraction(s: string): number | null {
	const trimmed = s.trim();
	if (!trimmed) return null;

	// Simple decimal
	const asNumber = Number.parseFloat(trimmed);
	if (Number.isFinite(asNumber) && /^[+-]?(\d+(\.\d*)?|\.\d+)$/.test(trimmed)) {
		return asNumber;
	}

	// Fractions: "8 1/2" or "8-1/2" or "17/2"
	let whole = 0;
	let fractionPart = trimmed;

	const spaceIndex = trimmed.indexOf(' ');
	const dashIndex = trimmed.indexOf('-');
	const separatorIndex = spaceIndex >= 0 ? spaceIndex : dashIndex;

	if (separatorIndex > 0) {
		const wholePart = trimmed.slice(0, separatorIndex).trim();
		if (!/^[+-]?\d+$/.test(wholePart)) return null;
		whole = Number.parseInt(wholePart, 10);
		fractionPart = trimmed.slice(separatorIndex + 1).trim();
	}

	const slashIndex = fractionPart.indexOf('/');
	if (slashIndex <= 0) return null;

	const numeratorStr = fractionPart.slice(0, slashIndex).trim();
	const denominatorStr = fractionPart.slice(slashIndex + 1).trim();

	if (!/^[+-]?\d+$/.test(numeratorStr) || !/^[+-]?\d+$/.test(denominatorStr)) return null;

	const numerator = Number.parseInt(numeratorStr, 10);
	const denominator = Number.parseInt(denominatorStr, 10);
	if (!Number.isFinite(numerator) || !Number.isFinite(denominator) || denominator === 0) return null;

	return whole + numerator / denominator;
}
